addons/io/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java

```diff
@@ -0,0 +1,910 @@
+                    endpointPoolConfig.setInterTransactionDelayMillis(DEFAULT_TCP_INTER_TRANSACTION_DELAY_MILLIS);
+                    endpointPoolConfig.setConnectMaxTries(Modbus.DEFAULT_RETRIES);
+                    return endpointPoolConfig;
+                }
+            });
+        });
+
+        GenericKeyedObjectPool<ModbusSlaveEndpoint, ModbusSlaveConnection> genericKeyedObjectPool = new ModbusConnectionPool(
+                connectionFactory);
+        genericKeyedObjectPool.setSwallowedExceptionListener(new SwallowedExceptionListener() {
+
+            @SuppressWarnings("null")
+            @Override
+            public void onSwallowException(@Nullable Exception e) {
+                LoggerFactory.getLogger(ModbusManagerImpl.class).error(
+                        "Connection pool swallowed unexpected exception: {}",
+                        Optional.ofNullable(e).map(ex -> ex.getMessage()).orElse("<null>"));
+            }
+        });
+        connectionPool = genericKeyedObjectPool;
+        this.connectionFactory = connectionFactory;
+    }
+
+    private Optional<ModbusSlaveConnection> borrowConnection(ModbusSlaveEndpoint endpoint) {
+        Optional<ModbusSlaveConnection> connection = Optional.empty();
+        KeyedObjectPool<ModbusSlaveEndpoint, ModbusSlaveConnection> pool = connectionPool;
+        if (pool == null) {
+            return connection;
+        }
+        long start = System.currentTimeMillis();
+        try {
+            connection = Optional.ofNullable(pool.borrowObject(endpoint));
+        } catch (Exception e) {
+            logger.warn("Error getting a new connection for endpoint {}. Error was: {} {}", endpoint,
+                    e.getClass().getName(), e.getMessage());
+        }
+        if (connection.isPresent()) {
+            ModbusSlaveConnection slaveConnection = connection.get();
+            if (!slaveConnection.isConnected()) {
+                logger.trace(
+                        "Received connection which is unconnected, preventing use by returning connection to pool.");
+                returnConnection(endpoint, connection);
+                connection = Optional.empty();
+            }
+        }
+        logger.trace("borrowing connection (got {}) for endpoint {} took {} ms", connection, endpoint,
+                System.currentTimeMillis() - start);
+        return connection;
+    }
+
+    private void invalidate(ModbusSlaveEndpoint endpoint, Optional<ModbusSlaveConnection> connection) {
+        KeyedObjectPool<ModbusSlaveEndpoint, ModbusSlaveConnection> pool = connectionPool;
+        if (pool == null) {
+            return;
+        }
+        connection.ifPresent(con -> {
+            try {
+                pool.invalidateObject(endpoint, con);
+            } catch (Exception e) {
+                logger.warn("Error invalidating connection in pool for endpoint {}. Error was: {} {}", endpoint,
```

Comment:
also here better leave the handling to the caller by throwing an exception

Reply:
I think leaving it to the caller might not be a good idea. After all, we would have to have `throws Exception` in the method signature, forcing the caller to `catch Exception`. 

With this method (`invalidate`), there are two different callers -- leading to similar/identical error handling.

Nothing you can do anyways in case "invalidation" fails...

Agree?

Reply:
As you like. If there isn't anything a caller can do about it or react in any special way, let's keep it as it is.